internal/app/readycheck.go

```diff
@@ -1,13 +1,161 @@
+	"github.com/gardener/etcd-wrapper/internal/util"
 
-}
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const (
+	// ReadyServerPort is the port number for the server that serves the readiness probe
+	ReadyServerPort       = int64(9095)
+	etcdConnectionTimeout = 5 * time.Second
+	etcdGetTimeout        = 5 * time.Second
+	etcdQueryInterval     = 2 * time.Second
+	etcdEndpointAddress   = ":2379"
+)
 
-// SetupReadinessProbe sets up the readiness probe for this application.
+var (
+	etcdReady bool
+)
+
+// Create a struct which will hold the last status for etcd.
+// In SetupReadinessProbe first call go a.queryAndUpdateEtcdReadiness()/queryAndUpdateEtcdReadiness this function will
+// periodically query etcd and updates the readiness struct. The handler just reads from the struct.
+// If the struct has no status it assumes NotAvailable else it returns the current status.
+
+// SetupReadinessProbe sets up the readiness probe for this application. It is a blocking function and therefore
+// the consumer should always call this within a go-routine unless the caller itself wants to block on this which is unlikely.
 func (a *Application) SetupReadinessProbe() {
-	//the caller should call this in a go-routine as this method will setup a HTTP router and call ListenAndServe which blocks.
-	//If the http server errors out then you will have to panic and that should cause the container to exit and then be restarted by kubelet.
+	// Start go routine to regularly query etcd to update its readiness status.
+	go a.queryAndUpdateEtcdReadiness()
+	// If the http server errors out then you will have to panic and that should cause the container to exit and then be restarted by kubelet.
+	if a.isTLSEnabled() {
+		http.Handle("/readyz", http.HandlerFunc(a.readinessHandler))
+		err := http.ListenAndServeTLS(fmt.Sprintf(":%d", ReadyServerPort), a.cfg.ClientTLSInfo.CertFile, a.cfg.ClientTLSInfo.KeyFile, nil)
+		if err != nil {
+			a.logger.Fatal("failed to start TLS readiness endpoint", zap.Error(err))
+		}
+	} else {
+		http.Handle("/readyz", http.HandlerFunc(a.readinessHandler))
+		err := http.ListenAndServe(fmt.Sprintf(":%d", ReadyServerPort), nil)
+		if err != nil {
+			a.logger.Fatal("failed to start readiness endpoint", zap.Error(err))
+		}
+	}
+}
+
+// queryAndUpdateEtcdReadiness periodically queries the etcd DB to check its readiness and updates the status
+// of the query into the etcdStatus struct. It stops querying when the application context is cancelled.
+func (a *Application) queryAndUpdateEtcdReadiness() {
+	// Create a ticker to periodically query etcd readiness
+	ticker := time.NewTicker(etcdQueryInterval)
+	defer ticker.Stop()
+
+	for {
+		// Query etcd readiness and update the status
+		etcdReady = a.isEtcdReady()
+
+		select {
+		// Stop querying and return when the context is cancelled
+		case <-a.ctx.Done():
+			a.logger.Error("stopped periodic DB query: context cancelled", zap.Error(a.ctx.Err()))
```

Comment:
I think we have to move `etcdReady` inside application struct.
Don't we have to add locks?

Reply:
Moved `etcdReady` and added a lock
PTAL

Reply:
any reason we have to move `etcdReady` as part of the `application` struct? And why do we need a mutex?

Reply:
> any reason we have to move etcdReady as part of the application struct?

Moving `etcdReady` into the Application struct would help encapsulate the state related to the application's readiness. This approach makes it easier to understand the relationship between the readiness state and the rest of the application logic. 
Why do we need `etcdReady` in pkg level scope? any other reason?

> readiness state is local to this file and is only used to construct a response for a readiness request that comes in on /readyz endpoint.

The `etcdReady` variable is being updated in one goroutine (`queryAndUpdateEtcdReadiness` function) and read in another `readinessHandler` function. This means that there is a possibility of a data race if these operations happen concurrently. Using a mutex ensures that the code is safe from data races, making it more robust and reliable.

Reply:
>The etcdReady variable is being updated in one goroutine (queryAndUpdateEtcdReadiness function) and read in another readinessHandler function.

There is currently no case of a data race. Let us discuss your concern offline.

Reply:
>Moving etcdReady into the Application struct would help encapsulate the state related to the application's readiness. This approach makes it easier to understand the relationship between the readiness state and the rest of the application logic.

There is no usage of ready state outside `readiness handler`. If and when we have that need it can be promoted to the application struct. Can you point to a single user outside this file?

Reply:
>Why do we need etcdReady in pkg level scope? any other reason?

There is no concept of file local variables in golang and therefore the next best is to have an unexposed variable which by default is package scoped.

Reply:
Ok me and @seshachalam-yv discussed and we agreed:
1. We will remove the need for a Mutex
2. We will move the ready state into Application struct